src/os/kqueue.rs

```diff
@@ -0,0 +1,329 @@
+    }
+}
+
+/// A filter that can be registered into a `kqueue`.
+pub trait Filter: FilterSealed {}
+
+unsafe impl<T: FilterSealed + ?Sized> FilterSealed for &T {
+    fn filter(&self, flags: FilterFlags, key: usize) -> libc::kevent {
+        (**self).filter(flags, key)
+    }
+}
+
+impl<T: Filter + ?Sized> Filter for &T {}
+
+/// Monitor this signal number.
+///
+/// No matter what `PollMode` is specified, this filter will always be
+/// oneshot-only.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Signal(c_int);
+
+impl Signal {
+    /// Monitor this signal number.
+    pub fn new(signal: c_int) -> Self {
+        Self(signal)
+    }
+
+    /// Get the signal number.
+    pub fn signal(self) -> c_int {
+        self.0
+    }
+}
+
+/// Alias for `libc::c_int`.
+#[allow(non_camel_case_types)]
+pub type c_int = i32;
+
+unsafe impl FilterSealed for Signal {
+    fn filter(&self, flags: FilterFlags, key: usize) -> libc::kevent {
+        libc::kevent {
+            ident: self.0 as _,
+            filter: libc::EVFILT_SIGNAL,
+            flags: flags | libc::EV_RECEIPT,
+            udata: key as _,
+            ..unsafe { mem::zeroed() }
+        }
+    }
+}
+
+impl Filter for Signal {}
+
+/// Monitor a child process.
+#[derive(Debug)]
+pub struct Process<'a> {
+    /// The child process to monitor.
+    child: &'a Child,
+
+    /// The operation to monitor.
+    ops: ProcessOps,
+}
+
+/// The operations that a monitored process can perform.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[non_exhaustive]
+pub enum ProcessOps {
+    /// The process exited.
+    Exit,
+
+    /// The process was forked.
+    Fork,
+
+    /// The process executed a new process.
+    Exec,
+}
+
+impl<'a> Process<'a> {
+    /// Monitor a child process.
+    pub fn new(child: &'a Child, ops: ProcessOps) -> Self {
+        Self { child, ops }
+    }
+}
+
+unsafe impl FilterSealed for Process<'_> {
+    fn filter(&self, flags: FilterFlags, key: usize) -> libc::kevent {
+        let fflags = match self.ops {
+            ProcessOps::Exit => libc::NOTE_EXIT,
+            ProcessOps::Fork => libc::NOTE_FORK,
+            ProcessOps::Exec => libc::NOTE_EXEC,
+        };
+
+        libc::kevent {
+            ident: self.child.id() as _,
+            filter: libc::EVFILT_PROC,
+            flags: flags | libc::EV_RECEIPT,
+            fflags,
+            udata: key as _,
+            ..unsafe { mem::zeroed() }
+        }
+    }
+}
+
+impl Filter for Process<'_> {}
+
+/// Wait for a timeout to expire.
+///
+/// Modifying the timeout after it has been added to the poller will reset it.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Timer {
+    /// Identifier for the timer.
+    id: usize,
+
+    /// The timeout to wait for.
+    timeout: Duration,
+}
+
+impl Timer {
+    /// Create a new timer.
+    pub fn new(id: usize, timeout: Duration) -> Self {
```

Comment:
if the fields of this struct would be marked `pub`, these methods could be omitted.